### DISCUSSION TITLE
powerdelta

### DISCUSSION THREAD

**Discussion by grericht** [arendst/Tasmota#18504] [Q&A] [answered]
Hello,
i have an esp8266 with a ttl ir read head as a smart meter interface. Everything works fine. I send the data to Home Assistant via MQTT. Unfortunately, this only works every 10 seconds. I have now read something about the powerdelta option and would like to use it. Unfortunately, this seems to be deactivated in the SMI? I get the following message:
18:11:13.244 CMD: powerdelta 30
18:11:13.253 MQT: tasmota/smi/stat/RESULT = {"Command":"Unknown"}

Why is this disabled? It makes a lot of sense with an SMI, doesn't it? My SmartMeter not only transmits the meter reading but also the current consumption. I would like to report a change much faster there.

Can I activate the function somehow?

**Comment by barbudor** [comment]
I think this is not supported as of today
ping @gemu2015

**Reply by grericht** [reply]
why? can i change the code?

**Reply by gemu2015** [reply, marked as answer]
SML does not support power delta. you may implement this with additional script code.
but if you only want to have values as fast as possible, simply add 16 to the decimal precision of the decode line. thus values are transmitted as they arrive from the meter. (see sml docs)

e.g change from
1,77070100010801ff@1000,Verbrauch_Tarif_1,kWh,Total_Tarif1,3
to
1,77070100010801ff@1000,Verbrauch_Tarif_1,kWh,Total_Tarif1,19

**Reply by grericht** [reply]
but powerdelta function would be also nice for sml! why it is not integrated?

**Reply by gemu2015** [reply]
because sml is a universal decoder and by using script code you can easily mimic everything else
powerdelta, voltagedelta, currentdelta, waterdelta, gasdelta etc.
thus you have much more flexibility

example

```
>D
pwr=0
delta=0


>S
; get sml decode line 1
pwr=sml[1]
; get delta value
delta=diff[pwr]

if delta>10 {
; up more than 10 units
->publish ....
}

if delta>-10 {
; down more than 10 units
->publish ....
}

if abs(delta)>10 {
; changed more than 10 units
->publish ....
}
```

**Reply by Qhilm** [reply]
So, I made some progress, I see the publish being triggered, but I am not formatting it right I think. Home Assistant is not adding the value to the right entity.

```
; get delta value in %
power=sml[2]
delta=diff[power]*100/power
; if abs(delta) > 5%, send value over MQTT
if abs(delta)>5 {
	=>Publish tele/%topic%/SENSOR {"Script":{"power":%power%}}
}
```

This is what I see in the tasmota console:

```
02:01:49.932 MQT: tele/tasmota_B2CA88/STATE = {"Time":"2024-02-19T02:01:49","Uptime":"0T07:33:08","UptimeSec":27188,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":4,"Objects":39},"Wifi":{"AP":1,"SSId":"REDACTED","BSSId":"REDACTED","Channel":11,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:03"}}
02:01:49.941 MQT: tele/tasmota_B2CA88/SENSOR = {"Time":"2024-02-19T02:01:49","DWS7412":{"energy":10278.7595,"power":240.26,"meter_id":"REDACTED"}}
02:02:49.943 MQT: tele/tasmota_B2CA88/STATE = {"Time":"2024-02-19T02:02:49","Uptime":"0T07:34:08","UptimeSec":27248,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":4,"Objects":39},"Wifi":{"AP":1,"SSId":"REDACTED","BSSId":"REDACTED","Channel":11,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:03"}}
02:02:49.952 MQT: tele/tasmota_B2CA88/SENSOR = {"Time":"2024-02-19T02:02:49","DWS7412":{"energy":10278.7635,"power":236.64,"meter_id":"REDACTED"}}
02:02:53.951 SCR: performs "Publish tele/tasmota_B2CA88/SENSOR {"Script":{"power":260.68}}"
02:02:53.956 MQT: tele/tasmota_B2CA88/SENSOR = {"Script":{"power":260.68}}
02:02:56.908 SCR: performs "Publish tele/tasmota_B2CA88/SENSOR {"Script":{"power":239.01}}"
02:02:56.913 MQT: tele/tasmota_B2CA88/SENSOR = {"Script":{"power":239.01}}
02:03:49.953 MQT: tele/tasmota_B2CA88/STATE = {"Time":"2024-02-19T02:03:49","Uptime":"0T07:35:08","UptimeSec":27308,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":4,"Objects":39},"Wifi":{"AP":1,"SSId":"REDACTED","BSSId":"REDACTED","Channel":11,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:03"}}
02:03:49.962 MQT: tele/tasmota_B2CA88/SENSOR = {"Time":"2024-02-19T02:03:49","DWS7412":{"energy":10278.7674,"power":234.04,"meter_id":"REDACTED"}}
02:04:49.915 MQT: tele/tasmota_B2CA88/STATE = {"Time":"2024-02-19T02:04:49","Uptime":"0T07:36:08","UptimeSec":27368,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":4,"Objects":39},"Wifi":{"AP":1,"SSId":"REDACTED","BSSId":"REDACTED","Channel":11,"Mode":"11n","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:03"}}
02:04:49.924 MQT: tele/tasmota_B2CA88/SENSOR = {"Time":"2024-02-19T02:04:49","DWS7412":{"energy":10278.7713,"power":231.91,"meter_id":"REDACTED"}}
```

I have the impression I need to publish everything the same way, including timestamp, so home assistant can properly associate it with the right entities...

**Reply by itsmicash** [reply]
> SML does not support power delta. you may implement this with additional script code. but if you only want to have values as fast as possible, simply add 16 to the decimal precision of the decode line. thus values are transmitted as they arrive from the meter. (see sml docs)
> 
> e.g change from 1,77070100010801ff@1000,Verbrauch_Tarif_1,kWh,Total_Tarif1,3 to 1,77070100010801ff@1000,Verbrauch_Tarif_1,kWh,Total_Tarif1,19

You‘re a life saver!

**Comment by grericht** [comment]
this is my script:

`>D
>B
=>sensor53 r
>M 1
+1,3,s,0,9600,SML
1,77070100010800ff@1000,Verbrauch_Summe,kWh,Verbrauch_Summe,7
1,77070100100700ff@1,Leistung_Summe,W,Watt_Summe,2
1,=h-- 
1,77070100240700ff@1,Leistung_L1,W,Watt_L1,2
1,77070100380700ff@1,Leistung_L2,W,Watt_L2,2
1,770701004c0700ff@1,Leistung_L3,W,Watt_L3,2
1,=h-- 
1,77070100200700ff@1,Spannung_L1,V,Volt_L1,1
1,77070100340700ff@1,Spannung_L2,V,Volt_L2,1
1,77070100480700ff@1,Spannung_L3,V,Volt_L3,1`

i should change to:

`>D
>B
=>sensor53 r
>M 1
+1,3,s,0,9600,SML
1,77070100010800ff@1000,Verbrauch_Summe,kWh,Verbrauch_Summe,7
1,77070100100700ff@1,Leistung_Summe,W,Watt_Summe,18
1,=h-- 
1,77070100240700ff@1,Leistung_L1,W,Watt_L1,18
1,77070100380700ff@1,Leistung_L2,W,Watt_L2,18
1,770701004c0700ff@1,Leistung_L3,W,Watt_L3,18
1,=h-- 
1,77070100200700ff@1,Spannung_L1,V,Volt_L1,1
1,77070100340700ff@1,Spannung_L2,V,Volt_L2,1
1,77070100480700ff@1,Spannung_L3,V,Volt_L3,1`

???

**Reply by gemu2015** [reply]
Yes

**Reply by grericht** [reply]
is there a wiki for that? why there was a 2 and why there are a 1 and a 2 and a 7 at the other numbers?

**Comment by gemu2015** [comment]
See sml docs
Decimal precision
